### PR TITLE
Allow sending JsonRpcRequests through wallet api

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Expose JSON-RPC methods through wallet-api.

## Changes

- Add a JSON-RPC client to the wallet-api.
- Updated piggybank to use wallet's JSON request.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: packages/browser-wallet-api-helpers/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## 0.1.2
```
Comment (orhoj) [inline packages/browser-wallet-api-helpers/CHANGELOG.md:3]: This should be 0.2.0

File: packages/browser-wallet-api-helpers/README.md [hunk 2]
```diff
@@ -114,3 +114,12 @@ const provider = await detectConcordiumProvider();
...
 provider.addChangeAccountListener((address) => (selectedAccountAddress = address));
 ```
+
+## Accessing node through JSON-RPC
+
+The wallet api exposes access to a JSON-RPC client. This allows a dApp to communicate with a node using JSON-RPC without knowing a server itself. The client can be accessed using the `node` field.
+
+```typescript
+const provider = await detectConcordiumProvider();
+const accountInfo = await provider.node.getAccountInfo(accountAddress);
```
Comment (orhoj) [inline packages/browser-wallet-api-helpers/README.md:124]: Accessing a class field directly seems a bit yucky. Perhaps it should be:
```
const accountInfo = await provider.getJsonRpcClient().getAccountInfo(accountAddress);
```

File: packages/browser-wallet-api-helpers/README.md [hunk 3]
```diff
@@ -114,3 +114,12 @@ const provider = await detectConcordiumProvider();
 let selectedAccountAddress: string | undefined = undefined;
 provider.addChangeAccountListener((address) => (selectedAccountAddress = address));
 ```
+
+## Accessing node through JSON-RPC
+
+The wallet api exposes access to a JSON-RPC client. This allows a dApp to communicate with a node using JSON-RPC without knowing a server itself. The client can be accessed using the `node` field.
```
Comment (orhoj) [inline packages/browser-wallet-api-helpers/README.md:120]:
```suggestion
The wallet API exposes access to a JSON-RPC client. This allows a dApp to communicate with the same node as the wallet is connected to, and enables dApps to access the JSON-RPC interface without being connected to a separate server itself. The client is accessed as shown in the example below.
```

File: packages/browser-wallet/src/background/index.ts [hunk 4]
```diff
@@ -4,11 +4,44 @@ import {
...
     MessageType,
 } from '@concordium/browser-wallet-message-hub';
-import { storedSelectedAccount, storedUrlWhitelist } from '@shared/storage/access';
+import { storedSelectedAccount, storedUrlWhitelist, storedJsonRpcUrl } from '@shared/storage/access';
 
+import { v4 as uuidv4 } from 'uuid';
 import bgMessageHandler from './message-handler';
 import { forwardToPopup, HandleMessage, HandleResponse, RunCondition, setPopupSize } from './window-management';
 
+// TODO Replace with concordiumSDK HttpProvider when cross-fetch supports service workers. https://github.com/lquixada/cross-fetch/issues/78
```
Comment (orhoj) [inline packages/browser-wallet/src/background/index.ts:13]: Could the HttpProvider be refactored to work in a service worker?

File: packages/browser-wallet/src/background/index.ts [hunk 5]
```diff
@@ -4,11 +4,44 @@ import {
...
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const jsonRpcRequest = async (url: string, method: string, params?: any) => {
```
Comment (orhoj) [inline packages/browser-wallet/src/background/index.ts:15]: Is `params` a `string`?

Comment (shjortConcordium) [reply packages/browser-wallet/src/background/index.ts]: Removed this.

File: packages/browser-wallet/src/background/index.ts (hunk 4 above)
Comment (shjortConcordium) [reply packages/browser-wallet/src/background/index.ts]: https://github.com/Concordium/concordium-node-sdk-js/pull/82

File: packages/browser-wallet-api-helpers/README.md [hunk 6]
```diff
@@ -147,3 +147,12 @@ provider.on('accountDisconnected', (accountAddress) => {
...
 provider.connect().then((accountAddress) => (selectedAccountAddress = accountAddress));
 ```
+
+## Accessing node through JSON-RPC
+
+The wallet API exposes access to a JSON-RPC client. This allows a dApp to communicate with the same node as the wallet is connected to, and enables dApps to access the JSON-RPC interface without being connected to a separate server itself. The client is accessed as shown in the example below.
+
+```typescript
+const provider = await detectConcordiumProvider();
+const accountInfo = await provider.getJsonRpcClient().getAccountInfo(accountAddress);
```
Comment (orhoj) [inline packages/browser-wallet-api-helpers/README.md:157]: I think the example should show how to get the client, and then it should also state that the client will be outdated if the network changes.

File: packages/browser-wallet/build/dev.ts [hunk 7]
```diff
@@ -32,6 +32,7 @@ const config: BuildOptions = {
     outdir: 'dist',
     define: {
         'process.env.NODE_ENV': `"${process.env.NODE_ENV}"`,
+        'document.baseURI': '""',
```
Comment (orhoj) [inline packages/browser-wallet/build/dev.ts:35]: This magic should be documented, or even better if we didn't need it.

File: packages/browser-wallet-api/src/wallet-api.ts [hunk 8]
```diff
@@ -92,6 +104,10 @@ class WalletApi extends EventEmitter implements IWalletApi {
     private handleEvent(type: EventType) {
         this.messageHandler.handleMessage(createEventTypeFilter(type), (msg) => this.emit(type, msg.payload));
     }
+
+    public getJsonRpcClient(): JsonRpcClient {
```
Comment (orhoj) [inline packages/browser-wallet-api/src/wallet-api.ts:108]: Is there access restriction on this? A non-whitelisted website should not be able to use it.

**Comment by orhoj** [comment]
If I don't have any account connected and I try to get the JSON-RPC client, then a connection request in the wallet is opened.

**Comment by orhoj** [comment]
The returned errors from the client should be better. I just get that the RPC call was rejected by the wallet, but my request was in fact just bad.

File: packages/browser-wallet-api-helpers/README.md [hunk 9]
```diff
@@ -147,3 +151,18 @@ provider.on('accountDisconnected', (accountAddress) => {
 // Connect to an account in the wallet again triggered elsewhere in the dApp.
 provider.connect().then((accountAddress) => (selectedAccountAddress = accountAddress));
 ```
+
+## Accessing node through JSON-RPC
+
+The wallet API exposes access to a JSON-RPC client. This allows a dApp to communicate with the same node as the wallet is connected to, and enables dApps to access the JSON-RPC interface without being connected to a separate server itself. The client is accessed as shown in the example below.
+The dApp does not need to recreate the client again when the wallet changes node or network, the client will always use the wallet's current connected JSON-RPC server.
+
+If you have not connected with the wallet (or previously been whitelisted) or if the user rejects signing the transaction, the commands will not be executed and the method will throw an error.
```
Comment (orhoj) [inline packages/browser-wallet-api-helpers/README.md:160]: If the user rejects signing the transactions? That does not seem relevant here.

Comment (shjortConcordium) [reply packages/browser-wallet-api-helpers/README.md]: Yes, it is not. Removed it.

File: packages/browser-wallet/src/background/index.ts [hunk 10]
```diff
@@ -4,11 +4,54 @@ import {
...
+                    // We lose the method's typing when sending the message.
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    method as any,
+                    params && JSONBig.parse(params)
+                )
+                .then(onSuccess)
+                .catch((e) => onFailure(e.toString()));
+        }
+    } else {
+        onFailure('RPC Call can only performed by whitelisted sites');
```
Comment (orhoj) [inline packages/browser-wallet/src/background/index.ts:51]:
```suggestion
        onFailure('RPC Call can only be performed by whitelisted sites');
```

**Review by orhoj — APPROVED** [review]

**Pull request merged**